File: class19/java/PES1903-HexagonalPorts.java
/*
 * PES1902 - Arquitetura Hexagonal (Ports & Adapters)
 * Aula 19: Camadas, MVC, Clean Architecture e Hexagonal
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Demonstra Ports & Adapters: o domínio define portas (interfaces),
 * os adaptadores implementam detalhes técnicos (DB, API).
 *
 * Compilar/executar:
 *   javac PES1902-HexagonalPorts.java && java HexagonalPorts
 */

import java.util.*;

// ===== Porta (interface no domínio) =====
interface RepositorioProduto {
    Produto buscarPorId(int id);
    void salvar(Produto produto);
}

// ===== Domínio (Core) =====
class Produto {
    int id;
    String nome;
    double preco;

    Produto(int id, String nome, double preco) {
        this.id = id; this.nome = nome; this.preco = preco;
    }

    @Override
    public String toString() {
        return String.format("Produto[%d] %s — R$ %.2f", id, nome, preco);
    }
}

class CatalogoService {
    private RepositorioProduto repositorio;

    CatalogoService(RepositorioProduto repositorio) {
        this.repositorio = repositorio;
    }

    Produto obterProduto(int id) {
        return repositorio.buscarPorId(id);
    }

    Produto cadastrarProduto(int id, String nome, double preco) {
        Produto p = new Produto(id, nome, preco);
        repositorio.salvar(p);
        return p;
    }
}

// ===== Adaptador de Memória =====
class RepositorioMemoria implements RepositorioProduto {
    private Map<Integer, Produto> dados = new HashMap<>();

    public Produto buscarPorId(int id) { return dados.get(id); }
    public void salvar(Produto produto) { dados.put(produto.id, produto); }
}

class HexagonalPorts {
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  ARQUITETURA HEXAGONAL — PORTS & ADAPTERS");
        System.out.println("=".repeat(60));

        RepositorioProduto repo = new RepositorioMemoria();
        CatalogoService catalogo = new CatalogoService(repo);

        catalogo.cadastrarProduto(1, "Notebook", 4500.00);
        catalogo.cadastrarProduto(2, "Mouse", 120.00);

        Produto p = catalogo.obterProduto(1);
        System.out.printf("%n  Produto encontrado: %s%n", p);

        System.out.println("\n✓ Domínio isolado: troque o adaptador sem alterar a lógica.");
        System.out.println("✓ Porta = interface; Adaptador = implementação concreta.");
    }
}

File: class19/python/PES1903-Hexagonal-Ports.py
"""
PES1902 - Arquitetura Hexagonal (Ports & Adapters)
Aula 19: Camadas, MVC, Clean Architecture e Hexagonal
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Demonstra Ports & Adapters: o domínio define portas (interfaces),
os adaptadores implementam detalhes técnicos (DB, API).
"""

from abc import ABC, abstractmethod


# ===== Porta (interface no domínio) =====
class RepositorioProduto(ABC):
    """Porta de saída: o domínio define O QUE precisa, não COMO."""

    @abstractmethod
    def buscar_por_id(self, id_produto):
        pass

    @abstractmethod
    def salvar(self, produto):
        pass


# ===== Domínio (Core) =====
class Produto:
    def __init__(self, id_produto, nome, preco):
        self.id = id_produto
        self.nome = nome
        self.preco = preco

    def __repr__(self):
        return f"Produto[{self.id}] {self.nome} — R$ {self.preco:.2f}"


class CatalogoService:
    """Serviço de domínio: depende da abstração (porta), não da implementação."""

    def __init__(self, repositorio: RepositorioProduto):
        self.repositorio = repositorio

    def obter_produto(self, id_produto):
        return self.repositorio.buscar_por_id(id_produto)

    def cadastrar_produto(self, id_produto, nome, preco):
        produto = Produto(id_produto, nome, preco)
        self.repositorio.salvar(produto)
        return produto


# ===== Adaptadores (implementações concretas) =====
class RepositorioMemoria(RepositorioProduto):
    """Adaptador: implementa a porta com armazenamento em memória."""

    def __init__(self):
        self._dados = {}

    def buscar_por_id(self, id_produto):
        return self._dados.get(id_produto)

    def salvar(self, produto):
        self._dados[produto.id] = produto


# ===== Aplicação (adaptador primário / driving) =====
def main_hexagonal():
    print("=" * 60)
    print("  ARQUITETURA HEXAGONAL — PORTS & ADAPTERS")
    print("=" * 60)

    # Injetar adaptador concreto na porta
    repo = RepositorioMemoria()
    catalogo = CatalogoService(repo)

    catalogo.cadastrar_produto(1, "Notebook", 4500.00)
    catalogo.cadastrar_produto(2, "Mouse", 120.00)

    p = catalogo.obter_produto(1)
    print(f"\n  Produto encontrado: {p}")

    print("\n✓ Domínio isolado: troque o adaptador sem alterar a lógica de negócio.")
    print("✓ Porta = interface; Adaptador = implementação concreta (DB, API, etc.).")


if __name__ == "__main__":
    main_hexagonal()
